P4 code generator for Tofino

taegyunkim opened the pull request:
This PR generates P4-14 code for Tofino by using the visitors generated by ANTLR for the .alu file format.

taegyunkim committed:
Simple stateful alu for tofino code generaion.

taegyunkim committed:
Remove unnecessary member functions and variables.

taegyunkim committed:
Dummy statelss alu visitor for tofino.

taegyunkim committed:
Initial pipelining for tofino code generator.

taegyunkim committed:
Intiial version of tofino code generation.

taegyunkim committed:
Removed Nested rule.

taegyunkim committed:
Add tofino stateful alu blackbox.

taegyunkim committed:
Canonicalize the names of variables.

taegyunkim committed:
Fix issues with parens.

taegyunkim committed:
Simplify tofino blackbox template.

taegyunkim committed:
Explicitly set to output_value.

taegyunkim committed:
Define a list of keywords and instruction dictinoary.

taegyunkim committed:
Change to output blackbox instead of sketch like code.

anirudhSK committed:
Add chip.p4 file from Pravein's repo.

anirudhSK committed:
Rename file.

anirudhSK committed:
Initial work towards turning Pravein's program into a template.

anirudhSK committed:
Generate a grid of ALUs.

anirudhSK committed:
Use "alu." as prefix for stateful+stateless ALU attributes.

anirudhSK committed:
Align attributes.

anirudhSK committed:
Prefix sketch_name. [skip ci]

taegyunkim committed:
Handle hole defs in a more appropriate way.

taegyunkim committed:
Rename alu_name -> alu_filename.

taegyunkim committed:
Use jinja2.StrictUndefined.

taegyunkim committed:
Move to templates directory.

taegyunkim committed:
Saner ifelseifelse stmt handling.

taegyunkim committed:
Connect stateless visitor with code generator.

taegyunkim committed:
Use filename instead of file, which could also mean Python's File objects.

taegyunkim committed:
Rename generators to visitors.

taegyunkim committed:
Update tofino stateless alu visitor to return simplified expressions.

taegyunkim committed:
Merge into tofino_p4.j2

taegyunkim committed:
Outputs what looks similar to tofino program.

taegyunkim committed:
Output result to p4 file.

taegyunkim committed:
Remove TofinoStatelessAluVisitor.

taegyunkim committed:
Use consistent field names.

taegyunkim committed:
Handle salu config holes.

taegyunkim committed:
Update comments.

taegyunkim committed:
Convert packet field definition parse rule to support arbitrary packet field names.

taegyunkim committed:
Convert state variable definition parse rule to support arbitrary state variable names.

anirudhSK reviewed (changes requested):
Please rebase against master and update generation of P4 code for stateless ALU (i.e., the file tofino_p4.j2) now that PR #7 has been merged into master.

taegyunkim committed:
Update tofino alu to use new grammar.

taegyunkim committed:
Update comments.

taegyunkim committed:
Print list of unused holes.

taegyunkim committed:
Rename generate* functions to eval*

anirudhSK committed:
Add template holes for opcode, operand1, operand2, and result.

anirudhSK committed:
Use subscript instead of dotted notation to access attributes.

It's clearer this way, even though jinja2 allows you to use either
(https://cewing.github.io/training.codefellows/assignments/day22/jinja2_walkthrough.html#item-and-attribute-access-in-templates)

anirudhSK committed:
Generate code for stateless ALU in P4.

anirudhSK committed:
Remove unused keys in stateful_alus dict.

anirudhSK committed:
Uncomment output_dst.

anirudhSK committed:
Add clarifying comment.

anirudhSK committed:
Add output_dst field.

anirudhSK committed:
Remove unused visitor method.

anirudhSK committed:
Generate P4 code for stateful operand mux holes.

anirudhSK committed:
Add missing semicolon.

anirudhSK committed:
Tofino compiler expects & (instead of &&); similarly for |.

anirudhSK committed:
We need to use double braces here.

anirudhSK committed:
Remove parenthesis in table.

anirudhSK committed:
Need a table suffix in apply statements.

anirudhSK committed:
Tofino compiler likes and, or, not (instead of &, |, ~).

anirudhSK committed:
Remove pragmas for now.

anirudhSK committed:
Simplify generated code a bit more before handing off to Tofino compiler.

anirudhSK committed:
Alternate, equivalent version of Tofino ALU that helps simplify generated P4 code.

anirudhSK committed:
Remove min, max from Tofino ALU because we don't know how to write P4 code for them yet.

anirudhSK committed:
Fix bug due to aliasing problem with lists of lists in Python

For example, see https://stackoverflow.com/questions/240178/list-of-lists-changes-reflected-across-sublists-unexpectedly

anirudhSK committed:
Ignore all table dependencies and use pragmas to specify placement;

Not sure if this will actually work.

anirudhSK committed:
Really simple example file that runs correctly on Tofino!

Use iterative_solver example_specs/simplest.sk example_alus/stateful_alus/tofino.alu example_alus/stateless_alus/stateless_alu_for_tofino.alu 1 1 '0,1,2,3' 10 --target-tofino

to generate P4 code. Then compile and run using the Tofino compiler and control-plane program.

anirudhSK committed:
Second program that works end to end on a 2-by-2 grid.

Need to remove bitwise operations because we are incorrectly translating them into P4.
For example, if x and y are ints, then x | y is being translated into (x == 1) || (y == 1),
which is incorrect.

taegyunkim committed:
Add support for immediate_operand.

taegyunkim committed:
Print unused holes before rendering and only when there is at least one.
